sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/Statistics.scala

```diff
@@ -49,12 +50,15 @@ object Statistics {
  * @param attributeStats Statistics for Attributes.
  * @param isRuntime Whether the statistics is inferred from query stage runtime statistics during
  *                  adaptive query execution.
+ * @param mapOutputStatistics the map output statistics from query stage runtime statistics during
+ *                            adaptive query execution.
  */
 case class Statistics(
     sizeInBytes: BigInt,
     rowCount: Option[BigInt] = None,
     attributeStats: AttributeMap[ColumnStat] = AttributeMap(Nil),
-    isRuntime: Boolean = false) {
+    isRuntime: Boolean = false,
+    mapOutputStatistics: Option[MapOutputStatistics] = None) {
```

Comment:
I feel it's a bit weird that `Statistics` has a field `MapOutputStatistics` where `MapOutputStatistics` is a physical shuffle operator only thing, but `Statistics` is for all logical operators. Maybe we can have:

```
RunTimeStatsSpec(
  isRuntime: Boolean,
  sizeInBytesPerPartition: Option[Array[Long]]
)

Statistics(
  runTimeStatsSpec: Option[RunTimeStatsSpec] = None,
  ...
)
```

Reply:
Have the similar thought and the change you point out seems a better approach. What do you think about ? @maropu @cloud-fan

Reply:
FYI, we took anther approach to support SHJ in AQE. We added a rule in `AdaptiveSparkPlanExec` to convert SMJ to SHJ according to shuffle stats, which requires no changes in `Statistics.scala` as the statistics is ready in `ShuffleStageInfo`.

The SMJ could also be converted to SHJ if applicable even if `PREFER_SORTMERGE` is set. cc @Liulietong

cc @luuliietong

Reply:
> We added a rule in AdaptiveSparkPlanExec to convert SMJ to SHJ according to shuffle stats

This looks like a better idea. Do you want to open a PR for it?